Stringable java classes support

dervan opened the pull request:
I've introduced Stringable classes support. In relation to that, I did some additional refactor with SchemaMapper - I think it significantly simplified schema classes handling in code.

dervan committed:
Stringable java classes support
Refactor with SchemaMapper

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastDeserializerGenerator.java:
```diff
@@ -3,41 +3,22 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+import com.sun.codemodel.*;
```
We don't want wildcarded import statements. Please fix your IDE not to wrap imports.

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastDeserializerGenerator.java:
```diff
@@ -63,106 +46,57 @@
+                        "Incorrect top-level writer schema: " + aliasedWriterSchema.getType());
             }
 
+            JCatchBlock catchBlock = tryDeserializeBlock._catch(codeModel.ref(Throwable.class));
```
I recommend removing this try block completely. We are already declaring in the method signature that we may throw an IOException, which in fact can be only produced by Decoder methods. In this case, any other exception different than IOException comes from badly generated code, let's not confuse this with IO related errors.

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastDeserializerGenerator.java:
```diff
@@ -308,173 +259,149 @@ private void processRecord(JVar schemaVariable, JVar containerVariable, final Sc
+                        getSchemaExpr(schema));
+            case FIXED:
+                JArray fixedBytesArray = JExpr.newArray(codeModel.BYTE);
+                for (char b : defaultValue.getTextValue().toCharArray()) {
```
I'm not sure if we can assume that we will always encounter text value here, I think that processing raw bytes would be safer. Or you can distinguish between bytes and text using isBinary on JsonNode.

dervan replied:
According to avro specification: "Default values for bytes and fixed fields are JSON strings, where Unicode code points 0-255 are mapped to unsigned 8-bit byte values 0-255."

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastDeserializerGenerator.java:
```diff
@@ -308,173 +259,149 @@ private void processRecord(JVar schemaVariable, JVar containerVariable, final Sc
+                return schemaMapper.getFixedValue(schema, fixedBytesArray, getSchemaExpr(schema));
+            case BYTES:
+                JArray bytesArray = JExpr.newArray(codeModel.BYTE);
+                for (byte b : defaultValue.getTextValue().getBytes()) {
```
As above.

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastDeserializerGenerator.java:
```diff
@@ -226,75 +147,105 @@ private void processRecord(JVar schemaVariable, JVar containerVariable, final Sc
         }
 
+        JMethod method = createMethod(recordWriterSchema, recordAction.getShouldRead());
+        method._throws(Throwable.class);
```
Why would we need to throw that? As above, isn't the IOException sufficient?

flowenol commented on src/main/java/com/rtbhouse/utils/avro/SchemaMapper.java:
```diff
@@ -0,0 +1,234 @@
+import java.util.List;
+import java.util.Map;
+
+class SchemaMapper {
```
I don't like the name of this class. It's more like a SchemaHelper, because it handles many schema-related tasks, not only schema to JClass mappings. Good refactor though, saves a lot of LOC in both serializer and deserializer.

flowenol commented on src/main/java/com/rtbhouse/utils/avro/SchemaMapper.java:
```diff
@@ -0,0 +1,234 @@
+        return schema.getProp("java-class") != null;
+    }
+
+    public static boolean haveStringableKey(Schema schema) {
```
You mean: hasStringableKey?

flowenol commented on src/main/java/com/rtbhouse/utils/avro/SchemaMapper.java:
```diff
@@ -0,0 +1,234 @@
+
+    public static boolean haveStringableKey(Schema schema) {
+        if (!Schema.Type.MAP.equals(schema.getType())) {
+            throw new SchemaMapperException("String schema expected!");
```
You mean: Map schema expected?

flowenol commented on src/main/java/com/rtbhouse/utils/avro/SchemaMapper.java:
```diff
@@ -0,0 +1,234 @@
+        if (!Schema.Type.STRING.equals(schema.getType())) {
+            throw new SchemaMapperException("String schema expected!");
+        }
+        return schema.getProp("java-class") != null;
```
SpecificData.CLASS_PROP - use this

flowenol commented on src/main/java/com/rtbhouse/utils/avro/SchemaMapper.java:
```diff
@@ -0,0 +1,234 @@
+        if (!Schema.Type.MAP.equals(schema.getType())) {
+            throw new SchemaMapperException("String schema expected!");
+        }
+        return schema.getProp("java-key-class") != null;
```
SpecificData.KEY_CLASS_PROP - use this

flowenol commented on src/main/java/com/rtbhouse/utils/avro/SchemaMapper.java:
```diff
@@ -0,0 +1,234 @@
+        }
+        return schema.getProp("java-key-class") != null;
+    }
+}
```
And what about stringable elements in arrays (SpecificData.ELEMENT_PROP)?

dervan replied:
They are handled by defining "java-class" property in "items" schema. Same method is described https://issues.apache.org/jira/browse/AVRO-1268 .
Note also, that it seems that SpecificData.ELEMENT_PROP is not used in SpecificDatumReader neither SpecificCompiler.java, so I guess that this is just for ReflectData use.

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastSerializerGenerator.java:
```diff
@@ -47,39 +46,27 @@ public FastSerializerGenerator(boolean useGenericTypes, Schema schema, File dest
+                throw new FastSerializerGeneratorException("Unsupported input schema type: " + schema.getType());
             }
+
+            serializeMethod.annotate(SuppressWarnings.class).param("value", "unchecked");
```
This suppression is really ugly and should be avoided. What causes such warnings? Can it be fixed in a proper way?

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastSerializerGenerator.java:
```diff
@@ -152,677 +153,186 @@ private void processArray(final Schema arraySchema, JVar containerVariable, JBlo
-        return primitiveClazz;
     }
 
     private boolean doesNotContainMethod(final Schema schema) {
```
Please sync the naming with deserializer - you renamed the similar method over there.

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastDeserializerGenerator.java:
```diff
@@ -872,497 +733,174 @@ private void processEnum(JVar containerVariable, JVar keyVariable, final Schema
+            return primitiveValueExpression;
         } else {
             body.directStatement(DECODER + "." + readFunction + ";");
+            return JExpr._null();
```
Why null in similar methods for Fixed and Enum and JExpr._null() here? Please explain or make it stricter.

flowenol commented on src/main/java/com/rtbhouse/utils/avro/SchemaMapper.java:
```diff
@@ -0,0 +1,234 @@
+    private final JCodeModel codeModel;
+    private final boolean useGenericTypes;
+
+    SchemaMapper(JCodeModel codeModel, boolean useGenericTypes) {
```
I do not encourage to use package access modifier, public is ok - this applies everywhere.

dervan committed:
Better handling Exception from Stringables

dervan commented:
I hope I  addressed your concerns. Necessary exceptions from stringable classes are handled more subtle now. Rest of requests are either implemented or explained.

dervan committed:
Stringable exceptions wrapped in AvroRuntimeException

flowenol commented on src/main/java/com/rtbhouse/utils/avro/FastDeserializerGenerator.java:
```diff
@@ -65,37 +82,45 @@
             JVar result = declareValueVar("result", aliasedWriterSchema, deserializeMethod.body());
 
-            JTryBlock tryDeserializeBlock = deserializeMethod.body()._try();
+            JBlock topLevelDeserializeBlock = new JBlock();
```
Consider following code:

```
            JMethod deserializeMethod = deserializerClass.method(JMod.PUBLIC, readerSchemaClass, "deserialize");

            JBlock topLevelDeserializeBlock = new JBlock();
            JVar result = declareValueVar("result", aliasedWriterSchema, topLevelDeserializeBlock);

            switch (aliasedWriterSchema.getType()) {
            case RECORD:
                processRecord(readerSchemaVar, result, aliasedWriterSchema, reader, topLevelDeserializeBlock,
                        fieldAction);
                break;
            case ARRAY:
                processArray(readerSchemaVar, result, null, aliasedWriterSchema, reader, topLevelDeserializeBlock,
                        fieldAction);
                break;
            case MAP:
                processMap(readerSchemaVar, result, null, aliasedWriterSchema, reader, topLevelDeserializeBlock,
                        fieldAction);
                break;
            default:
                throw new FastDeserializerGeneratorException(
                        "Incorrect top-level writer schema: " + aliasedWriterSchema.getType());
            }

            if (schemaAssistant.getExceptionsFromStringable().isEmpty()) {
                assignBodyBlock(deserializeMethod, topLevelDeserializeBlock);
            } else {
                JTryBlock tryBlock = deserializeMethod.body()._try();

                assignBodyBlock(tryBlock, topLevelDeserializeBlock);
                for (Class<? extends Exception> classException : schemaAssistant.getExceptionsFromStringable()) {
                    JCatchBlock catchBlock = tryBlock._catch(codeModel.ref(classException));
                    JVar exceptionVar = catchBlock.param("e");
                    catchBlock.body()._throw(JExpr._new(codeModel.ref(AvroRuntimeException.class)).arg(exceptionVar));
                }
            }

            deserializeMethod._throws(codeModel.ref(IOException.class));
            deserializeMethod.param(Decoder.class, DECODER);

            topLevelDeserializeBlock._return(result);
```

This allows omitting unwanted braces introduced by JBlock at topLevelDesrializeBlock.
assignBodyBlock in FastDeserializerGeneratorBase would look like this:

```
protected static void assignBodyBlock(Object statement, JBlock body) {
        try {
            Field field = statement.getClass().getDeclaredField("body");

            field.setAccessible(true);
            field.set(statement, body);
            field.setAccessible(false);

        } catch (ReflectiveOperationException e) {
            throw new FastDeserializerGeneratorException(e);
        }
    }
```

dervan committed:
Remove redundant parentheses

dervan committed:
Variables naming update and schema handlig improved

dervan committed:
More functional deserializer and casting cleanup
